src/typecheck/tff.ml

```diff
@@ -856,6 +856,14 @@ module Make
           { Ast.term = Ast.Symbol s; _ } as s_ast, l); _ } as ast ->
         parse_app env ast s s_ast l
 
+      | { Ast.term = Ast.App ({ Ast.term = Ast.Colon({Ast.term = Ast.Symbol s; _} as s_ast,expected);_}, l); _ } as ast ->
+        let t = match parse_app env ast s s_ast l with
+          | Term t -> t
+          | res -> _expected env "term" ast (Some res)
+        in
+        let ty = parse_ty env expected in
+        Term (T.ensure t ty)
```

Comment:
Hm... so this code as is can't really stay there: as pointed out in my first comment, it's not correct in general, and even your example feel... weird. I understand it's the only way to go around the arbitrary restrictions of the smtlib (which really failed the polymorphic datatype part imho), but considering it's for a special case not in the standard, it'd probably be best if that code could be in the user code if it is necessary (while being aware of the potential problems that could arise from such a code), but of course, the builtin mechanism is not expressive enough to allow that, :p
What I'm thinking is that the type-checker could offer a last resort escape hatch, i.e. at the end of the pattern-matching, instead or raising the error `Unhandled_ast`, the type-checker could defer to a user-provided function. The default function would raise the error as is currently done, but the user could provide her/his own function, that could include the code you  have here (of course, the Typer module would have to be extended to allow specifying such a function). Would that be good for your use-case ?

Reply:
If it's good for you, I'll make the change in the typechecker, ^^

Reply:
Ah, but then again, given the pattern, it would not end up being given to the escape hatch because it would be caught be the `App` pattern...

Reply:
Okay, so by adding two hooks in the type-checker, i.e. a `pre` hook  and a `post` hook, that would respectively be run as the first case of pattern matching and as the last case of the pattern matching, this can be implemented in user-land (for this sue case, only the pre-hook would have to be used, but since we're doing it, might as well do something complete).

Reply:
> I can change the production in the parser to add the colon at the right place; around the application.

I think this would be a bit too much of a hack (both in the implementation, and the expected semantics), and it would break the correct/legitimate use cases of the type annotation by applying it to an application that is not linked to the type annotation.
I'll implement the hook thing tomorrow so that this part of the code can also be move into colibri if that's okay with you.